kernel/src/scan/mod.rs

```diff
@@ -249,99 +227,87 @@ impl Scan {
-
-            let mut dv_mask = dv_treemap.map(treemap_to_bools);
-
-            for read_result in read_results {
-                let len = if let Ok(ref res) = read_result {
+            let read_results = engine
+                .get_parquet_handler()
+                .read_parquet_files(&[meta], global_state.read_schema.clone(), None)?;
+
+            let gs = global_state.clone();
+            Ok(read_results.into_iter().map(move |read_result| {
+                let read_result = read_result?;
+                // to transform the physical data into the correct logical form
+                let logical = transform_to_logical_internal(
+                    engine,
+                    read_result,
+                    &gs,
+                    &scan_file.partition_values,
+                    &self.all_fields,
+                    self.have_partition_cols,
+                );
+                let len = if let Ok(ref res) = logical {
                     res.length()
                 } else {
                     0
                 };
-
-                let read_result = match read_expression {
-                    Some(ref read_expression) => engine
-                        .get_expression_handler()
-                        .get_evaluator(
-                            self.physical_schema.clone(),
-                            read_expression.clone(),
-                            output_schema.clone(),
-                        )
-                        .evaluate(read_result?.as_ref()),
-                    None => {
-                        // if we don't have partition columns, the result is just what we read
-                        read_result
-                    }
-                };
-
                 // need to split the dv_mask. what's left in dv_mask covers this result, and rest
-                // will cover the following results
-                let rest = dv_mask.as_mut().map(|mask| mask.split_off(len));
-
-                let scan_result = ScanResult {
-                    raw_data: read_result,
-                    mask: dv_mask,
+                // will cover the following results. we `take()` out of `selection_vector` to avoid
+                // trying to return a captured variable. We're going to reassign `selection_vector`
+                // to `rest` in a moment anyway
+                let mut sv = selection_vector.take();
+                let rest = sv.as_mut().map(|mask| mask.split_off(len));
+                let result = ScanResult {
+                    raw_data: logical,
+                    mask: sv,
                 };
-                dv_mask = rest;
-                results.push(scan_result);
-            }
-        }
-        Ok(results)
+                selection_vector = rest;
+                Ok(result)
+            }))
+        }).flatten_ok().try_collect::<Result<ScanResult, crate::error::Error>, Vec<Result<ScanResult, crate::error::Error>>, crate::error::Error>()?.into_iter().collect()
```

Comment:
oh dear :( you could maybe save some space with a type alias or something or at the very least `Result<ScanResult, _>, Vec<Result<ScanResult, _>>, _>`

Reply:
Really, this is a rather excessive monadic chain inside a monadic chain:
```rust
scan_files.into_iter().map(|scan_file| { ... }).flatten_ok().try_collect::<...>()?.into_iter().collect()
```
We should probably break it up, because the code is not easy to grok right now -- neither for human nor compiler.

If I understand correctly:
1. Use the callback to build up a `Vec<ScanFile>` from the add actions
2. Convert each scan file into a `DeltaResult<FileDataReadResultIterator>`
3. Convert each successful file read result into an `Iterator<Item = DeltaResult<ScanResult>>`
4. Flatten the overall iterator of iterators into a `DeltaResult<Vec<ScanResult>>`

Which should mean something like:
```rust
// read each parquet file as 1+ scan results -- iterator of iterators -- and flatten it
let scan_files = scan_files.into_iter().map(|scan_file| {
    ... read_parquet_files(...) ...
}).flatten_ok();
// do everything else needed and turn it into a Result<Vec<_>>
scan_files.map(|scan_result| {
  let scan_result = scan_result?;
    ...
}).try_collect()?
```

Reply:
Yeah, I wanted to split it, but it's actually tricky. The `scan_file` is used again down in `transform_to_logical_internal`, so you can't `flatten_ok()` right away because you then lose the association with which `scan_file` we're transforming for. So you end up with something just as crazy anyway where you zip the results with the scan files and it's... messy.

BUT, turns out you just need to help the compiler understand the error type, and then we don't need the crazy turbofish, so hopefully how it is now is more acceptable.

Reply:
The new code is _much_ better, thanks!